arwen/contexts/async.go

```diff
@@ -0,0 +1,885 @@
+	storage := context.host.Storage()
+
+	storageKey := arwen.CustomStorageKey(arwen.AsyncDataPrefix, runtime.GetPrevTxHash())
+	data := storage.GetStorage(storageKey)
+	if len(data) == 0 {
+		return arwen.ErrNoStoredAsyncContextFound
+	}
+
+	loadedContext, err := context.deserialize(data)
+	if err != nil {
+		return err
+	}
+
+	context.callerAddr = loadedContext.callerAddr
+	context.returnData = loadedContext.returnData
+	context.asyncCallGroups = loadedContext.asyncCallGroups
+
+	return nil
+}
+
+// Delete deletes the persisted state of the AsyncContext from the contract storage.
+func (context *asyncContext) Delete() error {
+	runtime := context.host.Runtime()
+	storage := context.host.Storage()
+
+	storageKey := arwen.CustomStorageKey(arwen.AsyncDataPrefix, runtime.GetPrevTxHash())
+	_, err := storage.SetProtectedStorage(storageKey, nil)
+	return err
+}
+
+func (context *asyncContext) determineExecutionMode(destination []byte, data []byte) (arwen.AsyncCallExecutionMode, error) {
+	runtime := context.host.Runtime()
+	blockchain := context.host.Blockchain()
+
+	// If ArgParser cannot read the Data field, then this is neither a SC call,
+	// nor a built-in function call.
+	functionName, args, err := context.host.CallArgsParser().ParseData(string(data))
+	if err != nil {
+		return arwen.AsyncUnknown, err
+	}
+
+	sameShard := context.host.AreInSameShard(runtime.GetSCAddress(), destination)
+	if context.host.IsBuiltinFunctionName(functionName) {
+		if sameShard {
+			vmInput := runtime.GetVMInput()
+			isESDTTransfer, _, _ := context.isESDTTransferOnReturnDataFromFunctionAndArgs(runtime.GetSCAddress(), destination, functionName, args)
+			isAsyncCall := vmInput.CallType == vm.AsynchronousCall
+			isReturningCall := bytes.Equal(vmInput.CallerAddr, destination)
+
+			if isESDTTransfer && isAsyncCall && isReturningCall {
+				return arwen.ESDTTransferOnCallBack, nil
+			}
+
+			return arwen.AsyncBuiltinFuncIntraShard, nil
+		}
+
+		return arwen.AsyncBuiltinFuncCrossShard, nil
+	}
+
+	code, err := blockchain.GetCode(destination)
+	if len(code) > 0 && err == nil {
+		return arwen.SyncExecution, nil
+	}
+
+	return arwen.AsyncUnknown, nil
+}
+
+func (context *asyncContext) sendAsyncCallCrossShard(asyncCall *arwen.AsyncCall) error {
+	host := context.host
+	runtime := host.Runtime()
+	output := host.Output()
+
+	err := output.Transfer(
+		asyncCall.GetDestination(),
+		runtime.GetSCAddress(),
+		asyncCall.GetGasLimit(),
+		asyncCall.GetGasLocked(),
+		big.NewInt(0).SetBytes(asyncCall.GetValue()),
+		asyncCall.GetData(),
+		vm.AsynchronousCall,
+	)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// executeAsyncContextCallback will either execute a sync call (in-shard) to
+// the original caller by invoking its callback directly, or will dispatch a
+// cross-shard callback to it.
+func (context *asyncContext) executeContextCallback() error {
+	if !context.HasCallback() {
+		// TODO decide whether context.gasAccumulated should be restored here to
```

Comment:
I just realized that the implementation regarding context callbacks is wrong, because it conflates two concepts: 
* the final callback to be executed after all async callbacks have completed - we disabled this due to gas accumulation
* the callback _of the caller_, which `AsyncContext` is trying to execute in `PostprocessCrossShardCallback()`if the caller is in the same shard - this must not happen no matter what, because even if it were no problems with gas accumulation, the callback of the caller is already handled due to state stacking in `AsyncContext`.

This aspect of `AsyncContext` needs to be reworked.